lib/src/piece/piece.dart

```diff
@@ -55,6 +55,14 @@ abstract class Piece {
   /// Invokes [callback] on each piece contained in this piece.
   void forEachChild(void Function(Piece piece) callback);
 
+  /// The cost that this piece should apply to the solution when in [state].
+  ///
+  /// This is usually just the state's cost, but some pieces may want to tweak
+  /// the cost in certain circumstances.
+  // TODO(tall): Given that we have this API now, consider whether it makes
+  // sense to remove the cost field from State entirely.
+  int stateCost(State state) => state.cost;
```

Comment:
This API feels iffy to me. Is there any other way we can handle this? Would it work to do `if (_mustSplit) piece.pin(State(2, cost: _style.splitCost);` instead of `pin(State.split)`?

Reply:
> 1. That would work for this example, but I don't know if we'll run into cases in the future

I generally have a strong preference to lean on YAGNI and go for the least possible API surface until the other use case shows up.

I'm fine going forward with this as is and reconsidering the coupling between states, pieces, and the code writer after more of the language is implemented.

Reply:
If it's OK with you, I'm going to keep this as-is with a TODO comment still in there, and I'll think more about better ways to express relative priority between formatting choices and constraints between pieces. I have some ideas, but it's hard to evaluate them until we have some more complex pieces implemented that need them.